src/app/(pages)/codeLibrary/page.tsx

```diff
@@ -82,7 +91,123 @@ const CodeLibrary: React.FC = () => {
     { ssr: false },
   );
 
-  const ctx = useContext(DataContext);
+  // get the query data from the context
+  const selectedQuery = ctx?.selectedQuery;
+  const queryName = selectedQuery?.queryName ?? "query";
+
+  // ---------- checkbox management state --------- //
+  // Get custom value sets from context (assume structure: queryData.custom = { [valueSetId]: DibbsValueSet })
+  const [customCodeIds, setCustomCodeIds] = useState<{
+    [vsId: string]: DibbsValueSet;
+  }>({});
+  // initialize any existing custom value sets
+  useEffect(() => {
+    if (!selectedQuery?.queryId) return;
+    getSavedQueryById(selectedQuery.queryId).then(
+      (query: QueryTableResult | undefined) => {
+        // Only hydrate previous selections
+        const queryCustom =
+          (query?.queryData?.custom as {
+            [vsId: string]: DibbsValueSet & { includeValueSet?: boolean };
+          }) || {};
+        setCustomCodeIds(queryCustom);
+      },
+    );
+  }, [selectedQuery?.queryId]);
+
+  const handleValueSetToggle = (vsId: string, checked: boolean) => {
+    setCustomCodeIds((prev) => {
+      const vs = prev[vsId];
+      if (!vs) {
+        // Seed all concepts with the checked value
+        const valueSet = valueSets.find((vs) => vs.valueSetId === vsId);
+        if (!valueSet) return prev;
+        return {
+          ...prev,
```

Comment:
instead of spreading this information here, could we update the values directly on the `prev` dictionary since we have a key into that value? Something like

```ts
const valueSet = valueSets.find((vs) => vs.valueSetId === vsId);
valueSet.includeValueSet = checked
valueSet.concepts = valueSet.concepts.map((c) => ({...c, include: checked})
prev[vsId] = valueSet
return prev
```